### PR TITLE
Adding the Fol document and updating the Motr Design Doc List

### PR DESCRIPTION
**Pull request by VenkyOS**
Signed-off-by: Venkataraman Padmanabhan <venkataraman.padmanabhan@seagate.com>

### CODE REVIEW

File: doc/HLD_of_FOL.rst [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+- it is used by FDML to feed file system updates to fol consumers asynchronously;
+
+- more generally, a fol is used by various components (snapshots, addb, etc.) to consistently reconstruct the system state as at a certain moment in the (logical) past.
+
+Roughly speaking, a fol is a partially ordered collection of fol records, each corresponding to (part of) a consistent modification of file system state. A fol record contains information determining durability of the modification (how many volatile and persistent copies it has and where, etc.) and dependencies between modifications, among other things. When a client node has to modify a file system state to serve a system call from a user, it places a record it its (possibly volatile) fol. The record keeps track of operation state: has it been re-integrated to servers, has it been committed on the servers, etc. A server, on receiving a request to execute an update on a client behalf, inserts a record, describing the request into its fol. Eventually, fol is purged to reclaim storage, culling some of the records.
```
Comment (huanghua78) [inline doc/HLD_of_FOL.rst:31]: s
/it places a record it its (possibly volatile) fol
/it places a record in its (possibly volatile) fol

File: doc/HLD_of_FOL.rst [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+- an operation (or more generally, a piece of a file system state) is durable when it has enough persistent copies to survive any failure. Note that even as a record of a durable operation exists after a failure, the recovery might decide to undo the operation to preserve overall system consistency. Also note that the notion of failure is configuration dependent;
+
+- an operation is stable when it is guaranteed that the system state would be consistent with this operation having been executed. A stable operation is always durable. Additionally, M0 core guarantees that the recovery would never undo the operation. The system can renege stability guarantee only in the face of a catastrophe or a system failure;
+
+- updates U and V are conflicting or non-commutative if the final system state after U and W are executed depends on the relative order of their execution (note that system state includes information and result codes returned to the user applications); otherwise the updates are commutative. The later of two non-commutative updates depends on the earlier one. The earlier one is a pre-requisite of a later one (this ordering is well-defined due to serializability);
```
Comment (huanghua78) [inline doc/HLD_of_FOL.rst:51]: s
/if the final system state after U and W
/if the final system state after U and V

File: doc/HLD_of_FOL.rst [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+- auxiliary interfaces, used for fol pruning and querying.
+
+***********************
+Logical Specification
+***********************
+
+Overview
+=========
+
+Fol is stored in a transactional container1 populated with records indexed2 by lsn. An lsn is used to refer to a point in fol from other meta-data tables (epochs table, object index, sessions table, etc.). To make such references more flexible, a fol, in addition to genuine records corresponding to updates, might contain pseudo-records marking points on interest in the fol to which other file system tables might want to refer to (for example, an epoch boundary, a snapshot origin, a new server secret key, etc.). By abuse of terminology, such pseudo-records will be called fol records too. Similarly, as part of redo-recovery implementation, DTM might populate a node fol with records describing updates to be performed on other nodes.
```
Comment (huanghua78) [inline doc/HLD_of_FOL.rst:137]: Please note the first sentence:
```
Fol is stored in a transactional container1 populated with records indexed2 by lsn
```

There are two foot notes. Please arrange them in correct format.

File: doc/HLD_of_FOL.rst [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+Record Structure
+=================
+
+A fol record, added via the main fol interface, contains the following:
+
+1[R.BACK-END.TRANSACTIONAL] ST
+
+2[R.BACK-END.INDEXING] ST
```
Comment (huanghua78) [inline doc/HLD_of_FOL.rst:146]: These are the two foot notes.

File: doc/HLD_of_FOL.rst [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+Liveness and Pruning
+=====================
+
+A node fol must be prunable if only to function correctly on a node without persistent storage. At the same time, a variety of sub-systems both from M0 core and outside of it, might want to refer to fol records. To make pruning possible and flexible, each fol record is augmented with a reference counter, counting all outstanding references to the record. A record can be pruned iff its reference counter drops to 0 together with reference counters of all earlier (in lsn sense) unpruned records in the fol.
+
+Conformance
+=============
+
+- [r.fol.every-node]: on nodes with persistent storage, M0 core runs in the user space and the fol is stored in a data-base table. On a node without persistent storage, M0 core runs in the kernel space and the fol is stored in memory-only index. Data-base and memory-only index provide the same external interface, making fol code portable;
```
Comment (huanghua78) [inline doc/HLD_of_FOL.rst:174]: s
/On a node without persistent storage, M0 core runs in the kernel space and the fol is stored in memory-only index
/On a node without persistent storage, or M0 core runs in the kernel space, the fol is stored in memory-only index

File: doc/HLD_of_FOL.rst [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+- [r.fol.local-txn]: request handler inserts a record into FOL table in the context of the same transaction where update is executed. This guarantees WAL property of fol;
+
+- [R.FOL]: vacuous;
```
Comment (huanghua78) [inline doc/HLD_of_FOL.rst:178]: The title in the [] should use [CAPITAL.LETTER], just like those in other places.

All the word "fol" must be converted to upper case "FOL".

File: doc/HLD_of_FOL.rst (hunk 3 above)
Comment (huanghua78) [reply doc/HLD_of_FOL.rst:137]: the word "lsn" should be converted to "LSN".

Comment (huanghua78) [reply doc/HLD_of_FOL.rst:137]: "a fol" should be changed to "an FOL", just like "an LSN".

File: doc/HLD_of_FOL.rst (hunk 1 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:31]: Incorporated.

File: doc/HLD_of_FOL.rst (hunk 4 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:146]: Incorporated

File: doc/HLD_of_FOL.rst (hunk 6 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst]: Incorporated.

File: doc/HLD_of_FOL.rst [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+Fol is stored in a transactional container1 populated with records indexed2 by lsn. An lsn is used to refer to a point in fol from other meta-data tables (epochs table, object index, sessions table, etc.). To make such references more flexible, a fol, in addition to genuine records corresponding to updates, might contain pseudo-records marking points on interest in the fol to which other file system tables might want to refer to (for example, an epoch boundary, a snapshot origin, a new server secret key, etc.). By abuse of terminology, such pseudo-records will be called fol records too. Similarly, as part of redo-recovery implementation, DTM might populate a node fol with records describing updates to be performed on other nodes.
+
+Record Structure
+=================
+
+A fol record, added via the main fol interface, contains the following:
+
+[1][R.BACK-END.TRANSACTIONAL] ST
+
+[2][R.BACK-END.INDEXING] ST
```
Comment (huanghua78) [inline doc/HLD_of_FOL.rst:146]: Please move these two references to previous section: to the end of "Overview" section.
And 
modify:
Fol is stored in a transactional container1 populated with records indexed2 by lsn
to
Fol is stored in a transactional container[1] populated with records indexed[2] by lsn

**Review by huanghua78 — APPROVED** [review]

File: doc/HLD_of_FOL.rst (hunk 2 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:51]: Incorporated.

File: doc/HLD_of_FOL.rst (hunk 5 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst]: Incorporated.

File: doc/HLD_of_FOL.rst (hunk 7 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:141]: Incorporated.

File: doc/HLD_of_FOL.rst (hunk 3 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst]: Incorporated.

File: doc/HLD_of_FOL.rst [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+Liveness and Pruning
+=====================
+
+A node FOL must be prunable if only to function correctly on a node without persistent storage. At the same time, a variety of sub-systems both from M0 core and outside of it, might want to refer to FOL records. To make pruning possible and flexible, each FOL record is augmented with a reference counter, counting all outstanding references to the record. A record can be pruned if its reference counter drops to 0 together with reference counters of all earlier (in lsn sense) unpruned records in the fol.
+
+Conformance
+=============
+
+- [R.FOL.EVERY-NODE]: on nodes with persistent storage, M0 core runs in the user space and the FOL is stored in a data-base table. On a node without persistent storage, or M0 core runs in the kernel space, the fol is stored in memory-only index. Data-base and memory-only index provide the same external interface, making FOL code portable;
```
Comment (TechWriter-Mayur) [inline doc/HLD_of_FOL.rst:174]: Please capitalize the abbreviation fol to FOL.

File: doc/HLD_of_FOL.rst [hunk 9]
```diff
@@ -0,0 +1,290 @@
+============================================
+High level design of a File Operations Log
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:2]: Incorporated.

File: doc/HLD_of_FOL.rst [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+============================================
+
+This document presents a high level design (HLD) of a file operations log (a FOL) of Motr M0 core. The main purposes of this document are: (i) to be inspected by M0 architects and peer designers to ascertain that high level design is aligned with M0 architecture and other designs, and contains no defects, (ii) to be a source of material for Active Reviews of Intermediate Design (ARID) and detailed level design (DLD) of the same component, (iii) to serve as a design reference document.
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst]: Incorporated.

File: doc/HLD_of_FOL.rst [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+
+The intended audience of this document consists of M0 customers, architects, designers and developers.
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:13]: Deferred.

File: doc/HLD_of_FOL.rst [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+- [R.FOL.EVERY-NODE]: every node where M0 core is deployed maintains FOL;
+
+- [R.FOL.LOCAL-TXN]: a node FOL is used to implement local transactional containers
+
+- [R.FOL]: A File Operations Log is maintained by M0;
+
+- [R.FOL.VARIABILITY]: FOL supports various system configurations. FOL is maintained by every M0 back-end. FOL stores enough information to efficiently find modifications to the file system state that has to be propagated through the caching graph, and to construct network-optimal messages carrying these updates. A FOL can be maintained in volatile or persistent transactional storage;
+
+- [R.FOL.LSN]: A FOL record is identified by an LSN. There is a compact identifier (LSN, Log Sequence Number) with which a log record can be identified and efficiently located;
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:84]: Incorporated.

File: doc/HLD_of_FOL.rst (hunk 8 above)
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst:180]: Incorporated.

File: doc/HLD_of_FOL.rst [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+
+*************
+Introduction
+*************
+
+A FOL is a central M0 data-structure, maintained by every node where M0 core is deployed and serving multiple goals:
+
+- it is used by a node data-base component to implement local transactions through WAL logging;
+
+- it is used by DTM to implement distributed transactions. DTM uses fol for multiple purposes internally:
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst]: Incorporated.

File: doc/HLD_of_FOL.rst [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+
+The intended audience of this document consists of M0 customers, architects, designers and developers.
+
+*************
+Introduction
+*************
+
+A FOL is a central M0 data-structure, maintained by every node where M0 core is deployed and serving multiple goals:
+
+- it is used by a node data-base component to implement local transactions through WAL logging;
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst]: Incorporated.

File: doc/HLD_of_FOL.rst [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+
+- a FOL of a node is a sequence of records, each describing an update carried out on the node, together with information identifying operations these updates are parts of, file system objects that were updated and their versions, and containing enough data to undo or redo the updates and to determine operation dependencies;
+
+- a record in a node FOL is uniquely identified by a Log Sequence Number (LSN). Log sequence numbers have two crucial properties:
+
+  - a FOL record can be found efficiently (i.e., without FOL scanning) given its LSN, and
+
+  - for any pair of conflicting updates recorded in the FOL, the LSN of the pre-requisite is less than that of the dependent update (Note: clearly this property implies that lsn     data-type has infinite range and, hence, is unimplementable in practice. What is in fact required is that this property holds for any two conflicting updates sufficiently     close in logical time, where precise closeness condition is defined by the fol pruning algorithm. The same applies to object versions.);
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_FOL.rst]: Incorporated.

**Pull request merged**